Verifier: Add IBM Secure Execution driver framework

huoqifeng opened the pull request:
Fixes: #342 
This is kbs side code and related with PR: https://github.com/confidential-containers/guest-components/pull/492/
Depends on: 
 - [x] https://github.com/virtee/kbs-types/issues/26

# The IBM SE Remote Attestation flow:
<img width="1438" alt="image" src="https://github.com/confidential-containers/trustee/assets/15607193/8cd52762-2dbf-443a-b518-79ea4d57e7cd">

- The verifier generate the encrypted attestation-request based on hkd, CA, signing_key, a measurement key  and a nonce, the encrypted data is protected by a symmetric attestation request protection key,  which is encrypted using the Host-key document
- Verifier sends the request to attester
- Firmware on the Attester's system decrypts the request via private host-key and calculates the evidence based on the encrpted part of the request (Measurement key + nonce)
- Attester send the evidence to verifier
- Verifier recalculates the evidence based on the Configuration UID, Additional data, user-data, guest image hashes, and nonce. (its a HMAC-SHA512 with the measurement key as secret)
- if both HMACs, the one from the Firmware and the calculated one from the verifier match -> attestation success

Xynnn007 commented on kbs/src/api/src/attestation/mod.rs:
```diff
@@ -32,6 +32,9 @@ pub trait Attest: Send + Sync {
     async fn verify(&self, tee: Tee, nonce: &str, attestation: &str) -> Result<String>;
+
+    /// generate the challenge payload to pass to attester based on Tee and nonce
+    async fn generate_challenge(&self, tee: Tee, nonce: &str) -> Result<Challenge>;
```
Hi @huoqifeng , currently we will get the challenge nonce from KBS side [here](https://github.com/confidential-containers/trustee/blob/main/kbs/src/api/src/session.rs#L78). 

And we have decoupled the code. Now the challenge will be check binding by calling Attestation Service's [reportdata/runtimedata api parameter](https://github.com/confidential-containers/trustee/blob/main/kbs/src/api/src/attestation/coco/grpc.rs#L104C9-L104C100)

So I think what is needed to be done on the verifier side is only (If SE is similiar with TDX/SNP/CCA ...)
1. implement a verifier plugin of `attester`
2. add se to https://github.com/confidential-containers/trustee/blob/main/attestation-service/attestation-service/src/bin/grpc/mod.rs#L29-L41, https://github.com/confidential-containers/trustee/blob/main/attestation-service/attestation-service/src/bin/restful/mod.rs#L54-L69, https://github.com/confidential-containers/trustee/blob/main/attestation-service/protos/attestation.proto#L5-L15
3. Document update
(I might ignore something else that could be figured out due to CI error)

huoqifeng replied:
@Xynnn007 thanks for the good suggestions, there are some specific requirements for IBM SE, like:
- On kbs side, the `challenge` need include an `attestation request`.
like https://github.com/confidential-containers/trustee/blob/main/kbs/src/api/src/session.rs#L78, IBM SE has an `attestation request` need be generated by `verifier`, this request `must` be generated on verifier and pass to attester. We can use the `nonce` field but the meaning has been changed from a random str to something generated from verifier driver, alternative as I made in this draft PR, we can also use field `extra_params`, anyway, IBM SE need pass `attestation request` generated by verifier but not only random string.
- On guest side, this `attestation request` will be used to issue evidence
Like in https://github.com/confidential-containers/guest-components/blob/main/attestation-agent/kbs_protocol/src/client/rcar_client.rs#L184, we need `runtime_data` can be deserialized in `attester` driver, because IBM SE attester requires the `attestation request` passed from verifier side to generate the evidence. 

Lets talk about it this afternoon and make it clear. Thanks!

huoqifeng replied:
@Xynnn007 https://github.com/ibm-s390-linux/s390-tools/tree/master/pvattest, this is a reference for c approach, rust is in-developing.

huoqifeng commented:
@Xynnn007 @fitzthum I'd like start the review process of the API change to support IBM SE. May you please help on that?

Xynnn007 commented:
Oh. I am still worried about the breakage for the PR. As it couple the KBS logic and AS logic to prepare an `attreq.bin`. After some reading of https://www.ibm.com/docs/en/linuxonibm/pdf/l130se03.pdf, I think of some different design ways for this and I must ignore something.

#### Way 1: 

Con: We need to embed the host cert key chain (`host key doc`, `CA cert`, `IBM_sign_key_cert`) into the guest image

The logic of `attester` side:
1. After receiving `challenge`, generate an `arp.key` by `HKDF(challenge || host_key_doc || CA cert || IBM_sign_key_cert)`
2. Generate a report using the `arp.key` and other materials using `pvattest`

The logic of `verifier` side:
1. Use the same `host key doc`, `CA cert`, `IBM_sign_key_cert` and `challenge` to generate a same `arp.key`, and verify with `pvattest verify -i attresp.bin --arpk arp.key --hdr hdr.bin`

Rationality:
Looking at the entire attestation protocol of pvattest, the most critical thing is this `arp.key`, which is used for anti-replay, similar to the Nonce of other architectures, so we combine platform information `host_key_doc || intermediate CA certificate || IBM sign key cert` together with `challenge`.  The strong binding information is used as the derivation factor of this key to ensure that the attester cannot use an instance on another platform to forge evidence that can pass verification - because the `arp.key` of the verifier side must be based on the certificate of the target environment.

BTW, if SE guest could support initdata (either though `hostdata` mechanism or vTPM), we could inject those documents when launching the guest without embed them inside the guest.

#### Way 2: Change the `verifier` API

This is the current implementation.

Con: Bring extra `generate_challenge_extra_params()` for all verifiers and might couple the current design of `KBS` and `AS`.

This means every time KBS receives a RCAR request, it should first call the backend AS to get an `extra_params`. This API is somehow specific for CoCo hinting that Intel-TA and other attestation services might not support such.

#### Way 3: Integrite IBM SE verifier logic inside KBS, and attester logic in KBS-client

Con: No `se` plugin will be support in `attester` and `verifier` crate, and thus not be support by `AS`. `KBS` should store `arp.key`s and `se_guest.hdr`s

This is based on a different attestation protocol of SE. We can introduce some conditional judgement for SE on the both KBS and KBS client side.

#### Way 4: Integrite IBM SE verifier logic inside KBS but attester logic in `attester` crate

Con: `KBS` should store `arp.key`s and `se_guest.hdr`s.

On the attester side, we can reuse `report_data/challenge` field for SE to present the `host key doc`, `CA cert`, `IBM_sign_key_cert` and `attreq.bin` bundle and generate an `attresp.bin`

On the verifier side, `KBS` will implement the generation of `attreq.bin`, and the verification of `attresp.bin`.

huoqifeng commented:
> Oh. I am still worried about the breakage for the PR. As it couple the KBS logic and AS logic to prepare an `attreq.bin`. After some reading of https://www.ibm.com/docs/en/linuxonibm/pdf/l130se03.pdf, I think of some different design ways for this and I must ignore something.
> #### Way 1:
> 
> Con: We need to embed the host cert key chain (`host key doc`, `CA cert`, `IBM_sign_key_cert`) into the guest image
> 
> The logic of `attester` side:
> 
>     1. After receiving `challenge`, generate an `arp.key` by `HKDF(challenge || host_key_doc || CA cert || IBM_sign_key_cert)`
> 
>     2. Generate a report using the `arp.key` and other materials using `pvattest`
> 
> 
> The logic of `verifier` side:
> 
>     1. Use the same `host key doc`, `CA cert`, `IBM_sign_key_cert` and `challenge` to generate a same `arp.key`, and verify with `pvattest verify -i attresp.bin --arpk arp.key --hdr hdr.bin`
> 
> 
> Rationality: Looking at the entire attestation protocol of pvattest, the most critical thing is this `arp.key`, which is used for anti-replay, similar to the Nonce of other architectures, so we combine platform information `host_key_doc || intermediate CA certificate || IBM sign key cert` together with `challenge`. The strong binding information is used as the derivation factor of this key to ensure that the attester cannot use an instance on another platform to forge evidence that can pass verification - because the `arp.key` of the verifier side must be based on the certificate of the target environment.
> 
> BTW, if SE guest could support initdata (either though `hostdata` mechanism or vTPM), we could inject those documents when launching the guest without embed them inside the guest.
> #### Way 2: Add a `get_challenge` to the `verifier` API
> 
> This is the current implementation.
> 
> Con: `get_challenge()` might mean nothing to other platforms.
> 
> This means every time KBS receives a RCAR request, it should first call the backend AS to call `get_challenge()`.
> 
> This hints that CoCoAS should also support 4-pass protocol besides current 2-pass protocol, s.t. two ways to use CoCoAS
> 
>     1. `get_challenge()` and then [evaluate()](https://github.com/confidential-containers/trustee/blob/main/attestation-service/protos/attestation.proto#L92)
> 
>     2. directly [evaluate()](https://github.com/confidential-containers/trustee/blob/main/attestation-service/protos/attestation.proto#L92)
> 
> 
> cc @jialez0
> #### Way 3: Integrite IBM SE verifier logic inside KBS, and attester logic in KBS-client
> 
> Con: No `se` plugin will be support in `attester` and `verifier` crate, and thus not be support by `AS`. `KBS` should store `arp.key`s and `se_guest.hdr`s
> 
> This is based on a different attestation protocol of SE. We can introduce some conditional judgement for SE on the both KBS and KBS client side.
> #### Way 4: Integrite IBM SE verifier logic inside KBS but attester logic in `attester` crate
> 
> Con: `KBS` should store `arp.key`s and `se_guest.hdr`s.
> 
> On the attester side, we can reuse `report_data/challenge` field for SE to present the `host key doc`, `CA cert`, `IBM_sign_key_cert` and `attreq.bin` bundle and generate an `attresp.bin`
> 
> On the verifier side, `KBS` will implement the generation of `attreq.bin`, and the verification of `attresp.bin`.

Thanks @Xynnn007 for good suggestions, I think `way 2` approach is used in this PR, `generate_challenge_extra_params` is added in verifier with a default impplementation. I'm happy to chnage the function name. CC @jialez0

huoqifeng commented:
@Xynnn007 @jialez0 we also want to add the `extra-params` in `Attestation` like https://github.com/huoqifeng/kbs-types/blob/271ef4bac54035fb25fa44b24e5bbca844cd057c/src/lib.rs#L74-L75, how do you think?

fitzthum commented:
This is a tricky one. If we get this extra challenge information from the AS, does that mean that the AS will have to maintain some per-connection state?

huoqifeng commented:
> This is a tricky one. If we get this extra challenge information from the AS, does that mean that the AS will have to maintain some per-connection state?

@fitzthum No need keep the state. For IBM SE, the attester will embed the original encrypted attestation request in evidence also, the verifier will extract the request first and then perform the verifying.

Xynnn007 commented:
Well, as we might not change the low level of s390x attestation workflow involving pvattest, there are now some best ways for me as declared in https://github.com/confidential-containers/trustee/pull/345#issuecomment-1999175504

Way 1 if IBM SE cert chain for the guest will not change during the whole lifetime of the guest
Way 2 if we agree on a new API of get_nonce() to CoCoAS
Way 3 will surely achieve that s390x get resources after remote attestation, even if no code is added to attester and CoCo-AS

jialez0 commented on attestation-service/verifier/src/lib.rs:
```diff
@@ -152,6 +166,12 @@ pub trait Verifier {
         expected_init_data_hash: &InitDataHash,
     ) -> Result<TeeEvidenceParsedClaim>;
+
+    async fn generate_challenge_extra_params(
```
Do not use `extra_params` as function name here, because this is a conception of KBS protocol. I think `generate_challenge` is enough.

huoqifeng replied:
fixed

jialez0 commented on attestation-service/verifier/src/lib.rs:
```diff
@@ -152,6 +166,12 @@ pub trait Verifier {
     ) -> Result<TeeEvidenceParsedClaim>;
+
+    async fn generate_challenge_extra_params(
+        &self,
```
I think there may be some extra input parameters when call this function in the future, so can you remain a optional input parameter here? Such as `tee_parameters: Option<Vec<u8>>`.

huoqifeng replied:
fixed.

jialez0 commented on attestation-service/attestation-service/src/lib.rs:
```diff
@@ -239,6 +239,11 @@ impl AttestationService {
         self.rvps.verify_and_extract(message).await
     }
+
+    pub async fn generate_challenge_extra_params(&self, tee: Tee) -> Result<String> {
```
`generate_challenge_extra_params` -> `generate_challenge`

huoqifeng replied:
fixed

jialez0 commented on attestation-service/attestation-service/src/bin/restful/mod.rs:
Why don't you add generate challenge API in restful and grpc AS implementation? (What I can see is that new API is only in the AS-library)

huoqifeng replied:
Checking, will add it also. Thanks!

huoqifeng replied:
Fixed

jialez0 commented:
> Well, as we might not change the low level of s390x attestation workflow involving pvattest, there are now some best ways for me as declared in [#345 (comment)](https://github.com/confidential-containers/trustee/pull/345#issuecomment-1999175504)
> 
> Way 1 if IBM SE cert chain for the guest will not change during the whole lifetime of the guest Way 2 if we agree on a new API of get_nonce() to CoCoAS cc @jialez0 Way 3 will surely achieve that s390x get resources after remote attestation, even if no code is added to attester and CoCo-AS

@Xynnn007 I prefer way 2, I think AS is reasonable to provide this `generate_challenge` (or `get_nonce`) API. What's more, I think AS maintain a simple `Nonce Map` is not a very unreasonable thing, it is not a true "state save" like KBS sessions, and the Nonce Map here will be a very lightweight structure. cc @fitzthum

Xynnn007 commented:
> @Xynnn007 I prefer way 2, I think AS is reasonable to provide this `generate_challenge` (or `get_nonce`) API. What's more, I think AS maintain a simple `Nonce Map` is not a very unreasonable thing, it is not a true "state save" like KBS sessions, and the Nonce Map here will be a very lightweight structure. cc @fitzthum

Some more contexts are, as a separate attestation service, CoCoAS should have its own mechanism to prevent reply attack when the user does not use KBS.

So we add the new API `get_nonce()` and binds it to a same context of `attestation()` to let the AS check if the evidence is fresh enough. At the same time, only call `attestation()` to check an evidence is also supported.

fitzthum reviewed (commented):
A few comments. Sorry for the delay. I don't have a full picture without the actual attestation code, but I think this looks pretty good. I am open to adding support for the challenge generation.

I do wonder if there is a better name for `generate_challenge`. The current name makes it sound like somethign that every verifier would implement. Maybe `generate_supplemental_challenge` or something like that would be less confusing.

huoqifeng commented on attestation-service/verifier/src/se/mod.rs:
```diff
@@ -0,0 +1,76 @@
+        let hkds: Vec<String> = vec![String::new(); 2];
+        let certk = "cert_file_path";
+        let signk = "sign_file_path";
+        let arpk = "arpk_file_path";
```
@fitzthum yes, it's placeholder for now, will replace it when https://crates.io/crates/s390_pv is ready.

huoqifeng commented on attestation-service/verifier/src/se/mod.rs:
```diff
@@ -0,0 +1,76 @@
+    let claims_map = json!({
+        "serial_number": format!("{}", "SE-ID"),
+        "measurement": format!("{}", BASE64_STANDARD.encode(se.clone())),
+        "report_data": format!("{}", BASE64_STANDARD.encode(se.clone())),
```
Yes, I'll add this into doc once the crates is ready. Thanks! Tobin

huoqifeng commented:
> A few comments. Sorry for the delay. I don't have a full picture without the actual attestation code, but I think this looks pretty good. I am open to adding support for the challenge generation.
> 
> I do wonder if there is a better name for `generate_challenge`. The current name makes it sound like somethign that every verifier would implement. Maybe `generate_supplemental_challenge` or something like that would be less confusing.

Thanks @fitzthum , I'm open to the fn name. @Xynnn007 @jialez0 what's your suggestions on the fn name?

huoqifeng commented on kbs/src/api/src/attestation/mod.rs:
```diff
@@ -89,4 +93,25 @@ impl AttestationService {
+            }
+            #[cfg(feature = "intel-trust-authority-as")]
+            AttestationService::IntelTA(inner) => {
+                inner.generate_challenge(tee, tee_parameters).await
```
@fitzthum Shall we allow `kbs/src/api/src/attestation/intel_trust_authority/mod.rs` to decide error or a real value returned?

jialez0 commented:
> > A few comments. Sorry for the delay. I don't have a full picture without the actual attestation code, but I think this looks pretty good. I am open to adding support for the challenge generation.
> > I do wonder if there is a better name for `generate_challenge`. The current name makes it sound like somethign that every verifier would implement. Maybe `generate_supplemental_challenge` or something like that would be less confusing.
> 
> Thanks @fitzthum , I'm open to the fn name. @Xynnn007 @jialez0 what's your suggestions on the fn name?

I'm OK with `generate_supplemental_challenge`.

Xynnn007 commented on attestation-service/verifier/src/se/mod.rs:
```diff
@@ -0,0 +1,75 @@
+        let signk = "sign_file_path";
+        let arpk = "arpk_file_path";
+
+        let extra_params = attester.create(hkds, certk, signk, arpk)
```
```suggestion
        let challenge = attester.create(hkds, certk, signk, arpk)
```

The rename matches the function name to avoid confusion.

huoqifeng replied:
Good point, fixed.

Xynnn007 commented on attestation-service/verifier/src/se/mod.rs:
```diff
@@ -0,0 +1,75 @@
+
+    let claims_map = json!({
+        "serial_number": format!("{}", "SE-ID"),
+        "measurement": format!("{}", BASE64_STANDARD.encode(se.clone())),
```
Btw I'd suggest use `hex` rather than base64 for all claims. Wdyt ? @fitzthum @mkulke

Xynnn007 commented on attestation-service/verifier/src/se/seattest.rs:
```diff
@@ -0,0 +1,50 @@
+// Copyright (C) Copyright IBM Corp. 2024
```
Do we need this file? IMO things in `se/mod.rs` are enough.

huoqifeng replied:
Agreed, I'll drop this file later when the IBM SE crate ready.

Xynnn007 commented on kbs/src/api/src/attestation/coco/grpc.rs:
```diff
@@ -125,6 +126,27 @@ impl Attest for GrpcClientPool {
+    ) -> Result<String> {
+        let req = tonic::Request::new(ChallengeRequest {
+            tee: to_grpc_tee(tee).into(),
+            tee_params: String::from_utf8(tee_parameters.unwrap()).unwrap(),
```
We'd try to avoid using `unwrap()`s. See https://github.com/confidential-containers/trustee/blob/main/attestation-service/attestation-service/src/bin/grpc/mod.rs#L194-L195 for example

huoqifeng replied:
Fixed

Xynnn007 commented on kbs/src/api/src/attestation/mod.rs:
```diff
@@ -32,6 +32,10 @@ pub trait Attest: Send + Sync {
     async fn verify(&self, tee: Tee, nonce: &str, attestation: &str) -> Result<String>;
+
+    /// generate the challenge payload to pass to attester based on Tee and nonce
+    async fn generate_supplemental_challenge(&self, tee: Tee, tee_parameters: Option<Vec<u8>>)
```
As `extra_params` are defined as `String`, we can directly use `tee_parameters: Option<String>` here. helps to avoid `as_bytes` and `from_utf8` on both side.

huoqifeng replied:
Good point! fixed.

Xynnn007 commented on attestation-service/protos/attestation.proto:
```diff
@@ -88,8 +89,17 @@ message SetPolicyRequest {
 
+message ChallengeRequest {
+    Tee tee           = 1;
+    string tee_params = 2;
```
We might need some refactor for the `tee_params`. It is marked `string` here, but the source could be a JSON map for better organization. So `map<string, string>` might be better somehow.

BTW I do not see any `tee_params` generated from the `attester` side. Can you give a link?

huoqifeng replied:
Right, grpc and rest approach is new added in https://github.com/confidential-containers/trustee/pull/345/commits/3a6b84e8c8f6403a8801f5ed61ffd30d7f1bb3b0 according to @jialez0 's suggestion. IBM SE generates challenges directly.

Xynnn007 replied:
Recall this. See the comment in `attestation-service/attestation-service/src/bin/restful/mod.rs`

huoqifeng replied:
Fixed, thanks!

Xynnn007 reviewed (commented):
cool. Some comments..

huoqifeng commented:
@Xynnn007 @fitzthum @jialez0 addressed most of the comments, may you please have a look again? And we'll still need some time to get https://crates.io/crates/s390_pv ready, do you think it's OK to add the implementation details, claims details in a new PR? We focus the API change and framework in this PR?

Xynnn007 commented:
> @Xynnn007 @fitzthum @jialez0 addressed most of the comments, may you please have a look again? And we'll still need some time to get https://crates.io/crates/s390_pv ready, do you think it's OK to add the implementation details, claims details in a new PR? We focus the API change and framework in this PR?

Personally, it is a complete function, and the integration of API changes alone cannot make it work properly. Thus I prefer to wait until upstream is ready and then review and merge it together?

mkulke commented:
> Personally, it is a complete function, and the integration of API changes alone cannot make it work properly. Thus I prefer to wait until upstream is ready and then review and merge it together?

I agree. Also, from experience it's easier to merge attester + verifier together, since often you need to iterate on one of the components to make the full worklfow work (e.g. the evidence schema). So I'd recommend to start with a release on the external crate, then the attester module and finally the verifier (remember to bump the guest-components in this repo's kbs client). 

@huoqifeng You might also want to run an e2e test for the whole flow in `kbs/test` on an SE enclave. Not sure about the option to get an SE enclave as a runner, but even running it manually is probably.

I'll read the "Attestation Flow in IBM SE for Linux" document to understand better the need to extend the API for the SE challenge. before I can make an educated review.

Still, it would be good if you could describe in simple words what you are proposing and why the changes are required in the description of the PR.

huoqifeng commented:
> > Personally, it is a complete function, and the integration of API changes alone cannot make it work properly. Thus I prefer to wait until upstream is ready and then review and merge it together?
> 
> I agree. Also, from experience it's easier to merge attester + verifier together, since often you need to iterate on one of the components to make the full worklfow work (e.g. the evidence schema). So I'd recommend to start with a release on the external crate, then the attester module and finally the verifier (remember to bump the guest-components in this repo's kbs client).
> 
> @huoqifeng You might also want to run an e2e test for the whole flow in `kbs/test` on an SE enclave. Not sure about the option to get an SE enclave as a runner, but even running it manually is probably.
> 
> I'll read the "Attestation Flow in IBM SE for Linux" document to understand better the need to extend the API for the SE challenge. before I can make an educated review.
> 
> Still, it would be good if you could describe in simple words what you are proposing and why the changes are required in the description of the PR.

@mkulke @Xynnn007 @fitzthum @jialez0 I updated the description in the PR to explain the IBM SE Attestation flow and you can check it an see why a `generate_supplemental_challengs` is required in IBM SE case. I think which is also an option for other TEE in future.

huoqifeng commented:
> Does the Verifier (attestation-service in our case) have access to key material that would not be available of the Attester (attestation-agent in our case)?

The `a measurement key` is generated by verifier and will be passed to attester in encrypted format. This key is only visible in hardware/firmware but not in attester's user space.

> would it make sense that the attester creates the "attestion request" using the Challenge retrieved from the Verifier as nonce?
No, because of the above logic, the request can only be created and encrypted by verifier. I think this is also the philosophy described in RATS.

mkulke commented on attestation-service/verifier/src/lib.rs:
```diff
@@ -152,6 +166,13 @@ pub trait Verifier {
         expected_init_data_hash: &InitDataHash,
     ) -> Result<TeeEvidenceParsedClaim>;
+
+    async fn generate_supplemental_challenge(
```
we probably want to add api docs for this trait fn

huoqifeng replied:
Added, thanks!

mkulke commented on kbs/src/api/src/attestation/coco/grpc.rs:
```diff
@@ -125,6 +126,27 @@ impl Attest for GrpcClientPool {
+    ) -> Result<String> {
+        let req = tonic::Request::new(ChallengeRequest {
+            tee: to_grpc_tee(tee).into(),
+            tee_params: tee_parameters.unwrap(),
```
we shouldn't use unwrap/panic in library code. you can turn a None option into a Result using `Option::ok_or()`.

Do we have to use Option for `tee_parameters`? on kbs_types' Request type, the extra_params propery is [not an option](https://github.com/virtee/kbs-types/blob/3f70b78f18e39cfcd3c5225f56b49117c95ee79b/src/lib.rs#L48)

huoqifeng replied:
Sounds good to me, will fix it.

mkulke commented:
> @mkulke hope it makes sense.

@huoqifeng thanks for the details, I understand it better, probably not fully yet. It looks like the SE Attestation scheme requires a special attestation challenge to be set by the SE Verifier. The SE Attester will need it to create an evidence (while other attesters are able to consume a non-structured byte array).

huoqifeng commented:
> > @mkulke hope it makes sense.
> 
> @huoqifeng thanks for the details, I understand it better, probably not fully yet. It looks like the SE Attestation scheme requires a special attestation challenge to be set by the SE Verifier. The SE Attester will need it to create an evidence (while other attesters are able to consume a non-structured byte array).

Right, IBM SE verifier actually verify the evidence and returns true or false only, but we do inject information for further checking, like: Additional data, user-data, guest image hashes mentioned in the flow. I'll add the details when the crate is ready.

huoqifeng commented:
@fitzthum @Xynnn007 @jialez0 @mkulke , the IBM SE implementation has been added, would you please check it again?

Xynnn007 commented on Cargo.toml:
```diff
@@ -30,13 +30,14 @@ clap = { version = "4", features = ["derive"] }
 regorus = { version = "0.1.5", default-features = false, features = ["regex", "base64", "time"] }
 rstest = "0.18.1"
 serde = { version = "1.0", features = ["derive"] }
+serde_with = { version = "1.11.0", features = ["base64"] }
```
Let's put `serde_with` after `serde_json` to follow dictionary order

huoqifeng replied:
Fixed, thanks!

Xynnn007 commented on Cargo.toml:
```diff
@@ -47,3 +48,4 @@ tokio = { version = "1.23.0", features = ["full"] }
 tempfile = "3.4.0"
 tonic = "0.8.1"
 tonic-build = "0.8.0"
+s390_pv = "0.10.0"
```
Let's move `s390_pv` to `attestation-service/verifier` crate's `Cargo.toml`

huoqifeng replied:
Fixed, thanks!

Xynnn007 commented on attestation-service/attestation-service/src/bin/restful/mod.rs:
```diff
@@ -44,6 +44,12 @@ pub struct AttestationRequest {
     policy_ids: Vec<String>,
 }
 
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ChallengeRequest {
+    tee: String,
+    tee_params: String,
+}
```
BTW, we might need to refactor this request struct. As we might have more abstract contexts apart from `tee` and `tee_params`, like
```rust
#[derive(Debug, Serialize, Deserialize)]
pub struct ChallengeRequest {
    #[flatten]
    inner: HashMap<String, String>,
}
```

huoqifeng replied:
Fixed, thanks!

Xynnn007 commented on attestation-service/docs/parsed_claims.md:
```diff
@@ -93,6 +93,13 @@ The claim inherit the fields from the SEV-SNP claim with and additional `tpm` hi
+- `se.hdr.tag`: SE Header Tag (seht)
+- `se.image.phkh`: SE image Public host key hash
+- `se.attestation.phkh`: SE attestation Public host key hash
+- `se.user_data`: Custom attestation key owner data.
```
Nice to see the parsed claims here. Need to add a new empty line here.

huoqifeng replied:
Fixed, thanks!

Xynnn007 commented on attestation-service/verifier/src/lib.rs:
```diff
@@ -152,6 +166,21 @@ pub trait Verifier {
+    /// generate the evidence
+    ///
+    /// A optional `tee_parameters` comes from the attester side as the input.
+
```
If we want to use inline comment, we could delete extra empty lines between `///` lines and function definitions.

huoqifeng replied:
Fixed, thanks!

Xynnn007 commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,318 @@
+use std::{fs::File, io::Read};
+
+fn encrypt_measurement_key(key: &[u8], rsa_public_key: &PKey<Public>) -> Vec<u8> {
+    let mut encrypter = Encrypter::new(rsa_public_key).unwrap();
```
Let's try to avoid all the `unwrap()`s and replace with `anyhow` error handling.

huoqifeng replied:
Fixed, thanks!

Xynnn007 commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,318 @@
+}
+
+fn decrypt_measurement_key(key: &[u8], rsa_private_key: &PKey<Private>) -> Vec<u8> {
+    let mut decrypter = Decrypter::new(rsa_private_key).unwrap();
```
Let's try to avoid all the `unwrap()`s and replace with `anyhow` error handling.

Xynnn007 commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,318 @@
+}
+
+impl SeAttestationRequest {
+    pub fn from_slice(request: &[u8]) -> Result<Self> {
+        Ok(serde_json::from_slice(request).unwrap())
+    }
+
+    pub fn from_string(request: &str) -> Result<Self> {
+        Ok(serde_json::from_str(request).unwrap())
+    }
```
we have `serde_json::from_str` and `serde_json::from_slice` already ,thus no need to wrap the function again here.

huoqifeng replied:
Fixed, thanks!

Xynnn007 commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,318 @@
+        Ok(serde_json::from_str(request).unwrap())
+    }
+
+    pub fn create(
+        measurement: &[u8],
+        additional_data: &[u8],
+        user_data: &[u8],
+        cuid: &ConfigUid,
+        encr_measurement_key: &[u8],
+        encr_request_nonce: &[u8],
+        image_hdr_tags: &BootHdrTags,
+    ) -> Result<Self> {
+        Ok(Self {
+            measurement: measurement.to_vec(),
+            additional_data: additional_data.to_vec(),
+            user_data: user_data.to_vec(),
+            cuid: *cuid,
+            encr_measurement_key: encr_measurement_key.to_vec(),
+            encr_request_nonce: encr_request_nonce.to_vec(),
+            image_hdr_tags: *image_hdr_tags,
+        })
+    }
```
If this function is only to specify every field of the `SeAttestationResponse`, we could directly use
```rust
let a = SeAttestationResponse {
...
}
```

rather than define it again.

huoqifeng replied:
Fixed, thanks!

Xynnn007 commented on attestation-service/verifier/src/se/mod.rs:
```diff
@@ -0,0 +1,89 @@
+
+pub mod ibmse;
+
+const SE_HOST_KEY_DOCUMENTS_ROOT: &str = "/run/confidential-containers/ibmse/hkds";
+const SE_CERTIFICATES_ROOT: &str = "/run/confidential-containers/ibmse/certs";
+const SE_CERTIFICATE_ROOT_CA: &str = "/run/confidential-containers/ibmse/certs/ca";
+const SE_CERTIFICATE_REVOCATION_LISTS_ROOT: &str = "/run/confidential-containers/ibmse/crls";
+const SE_IMAGE_HEADER_FILE: &str = "/run/confidential-containers/ibmse/hdr/hdr.bin";
+const SE_MEASUREMENT_ENCR_KEY_PRIVATE: &str =
+    "/run/confidential-containers/ibmse/rsa/encrypt_key.pem";
+const SE_MEASUREMENT_ENCR_KEY_PUBLIC: &str =
```
Keep in mind that these files are set-up on the trustee side -- Trustee runs on a trusted place and does not need to be in TEE. So we'd better make the files configurable. `/run` might not be a good place.

huoqifeng replied:
Fixed, thanks!

Xynnn007 commented on kbs/src/api/src/attestation/coco/grpc.rs:
```diff
@@ -118,6 +134,28 @@ impl Attest for GrpcClientPool {
+        tee: Tee,
+        tee_parameters: String,
+    ) -> Result<String> {
+        let grpc_tee = to_grpc_tee(tee).unwrap();
```
We can use
```
let tee = serde_json::to_string(&tee)
            .context("CoCo AS client: serialize tee type failed.")?
            .trim_end_matches('"')
            .trim_start_matches('"')
            .to_string();
```
similiarly to what we use in `verify()` without defining `to_grpc_tee()`.

Xynnn007 reviewed (commented):
Good. We are getting there.

mkulke commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,318 @@
+    }
+
+    pub fn verify(&self, priv_key_file: &str) -> Result<SeAttestationClaims> {
+        let mut file = File::open(priv_key_file)?;
+        let mut contents = vec![];
+        file.read_to_end(&mut contents)?;
```
Does it make sense to re-read and re-process the file in every verify operation or could the private key be initialized once?

background: the top-level fns in the verifiers are async and we shouldn't perform costly sync code in those. There are async fs operations, but maybe we don't need it if we parse the key once.

huoqifeng replied:
OK, I'll use lazy load and cache in memory to do that. Thanks!

mkulke commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,318 @@
+        let meas_key = decrypt_measurement_key(&self.encr_measurement_key, rsa_private_key);
+        let nonce = decrypt_measurement_key(&self.encr_request_nonce, rsa_private_key);
+
+        if nonce.len() != 16 {
+            return Err(anyhow!("The nonce vector must have exactly 16 elements."));
+        }
+        let boxed_slice: Box<[u8]> = nonce.into_boxed_slice();
+        let boxed_array: Box<[u8; 16]> = match boxed_slice.try_into() {
+            Ok(ba) => ba,
+            Err(_) => return Err(anyhow!("Failed to convert nonce from Vec<u8> to [u8; 16].")),
+        };
+        let nonce_array: [u8; 16] = *boxed_array;
```
decrypt_measurement_key returns a Vec<u8>, so we can do just that, no?

```rust
let nonce_array: [u8; 16] = nonce
  .try_into()
  .ok_or(anyhow!("Failed to convert nonce from Vec<u8> to [u8; 16]."))?;
```

huoqifeng replied:
Fixed, thanks!

mkulke commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,318 @@
+
+        let measurement =
+            AttestationMeasurement::calculate(items, AttestationMeasAlg::HmacSha512, meas_key)
+                .unwrap();
```
the unwrap should be handled

huoqifeng replied:
Fixed, thanks!

mkulke commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,318 @@
+            debug!("Recieved: {:?}", self.measurement);
+            debug!("Calculated: {:?}", measurement.as_ref());
+            warn!("Attestation measurement verification failed. Calculated and received attestation measurement are not equal.");
+            return Err(anyhow!("Failed to verify the measurement!"));
```
Err(anyhow!(...) can be shortened to bail!

huoqifeng replied:
Fixed, thanks!

mkulke commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,318 @@
+        &mut image_hdr_tags,
+        pub_key_file,
+    )
+    .unwrap();
```
The unwrap should be handled

huoqifeng replied:
Fixed, thanks!

mkulke commented on kbs/src/api/src/http/attest.rs:
```diff
@@ -17,11 +18,17 @@ pub(crate) async fn auth(
+    let extra_params = attestation_service
+        .generate_supplemental_challenge(request.tee, request.extra_params.clone())
+        .await
+        .unwrap();
```
the unwrap should be handled

huoqifeng replied:
Fixed, thanks!

mkulke reviewed (commented):
thx. a couple of comments, mostly related to panics/unwrap.

huoqifeng commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,318 @@
+                warn!("The host key document in '{hkd}' contains more than one certificate!")
+            }
+            // Panic: len is == 1 -> unwrap will succeed/not panic
+            let c = certs.first().unwrap();
```
Fixed, thanks!

huoqifeng commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,318 @@
+        let encr_measurement_key =
+            encrypt_measurement_key(conf_data.measurement_key(), rsa_public_key);
+        let encr_request_nonce =
+            encrypt_measurement_key(conf_data.nonce().clone().unwrap().value(), rsa_public_key);
```
Fixed, thanks!

huoqifeng commented:
@mkulke @Xynnn007 thanks for the review and good suggestions! I have updated it in new commit, would you please check it again?

mkulke commented on attestation-service/attestation-service/src/bin/grpc/mod.rs:
```diff
@@ -195,6 +199,43 @@ impl AttestationService for Arc<RwLock<AttestationServer>> {
+            .inner
+            .get("tee_params")
+            .map_or(Err(Status::aborted("Error parse inner_tee tee_params")), Ok)?;
+        let tee = to_kbs_tee(&inner_tee).map_or(
+            Err(Status::aborted(format!(
+                "Error parse TEE type: {inner_tee}"
+            ))),
+            Ok,
+        )?;
```
```suggestion
        let tee = to_kbs_tee(&inner_tee).ok_or(Status::aborted(format!("Error parse TEE type: {inner_tee}")))?;
```

mkulke commented on attestation-service/attestation-service/src/bin/restful/mod.rs:
```diff
@@ -44,6 +44,14 @@ pub struct AttestationRequest {
 
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ChallengeRequest {
+    // tee: String,
+    // tee_params: String,
```
leftovers?

mkulke commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,302 @@
+}
+
+fn get_cached_file_or_read(filename: &str, content_ref: &Mutex<Option<Vec<u8>>>) -> io::Result<Vec<u8>> {
+    let mut guard = content_ref.lock().unwrap();
```
let's avoid the unwrap

mkulke commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,302 @@
+
+#[serde_as]
+#[derive(Debug, Serialize, Deserialize)]
+pub struct SeAttestationClaims {
```
for those structs, do you need to set the memory layout to `#[repr(c)]` maybe? If they originate from c libraries rust/llvm might decide to reorder/pad the fields.

mkulke commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,302 @@
+            bail!("Failed to verify the measurement!");
+        }
+        
+        // let userdata = serde_json::from_slice(&self.user_data)?;
+        // debug!("user_data: {:?}", userdata);
```
leftover?

huoqifeng commented on attestation-service/attestation-service/Cargo.toml:
```diff
@@ -4,7 +4,7 @@ version = "0.1.0"
 edition = "2021"
 
 [features]
-default = [ "restful-bin", "rvps-grpc", "rvps-builtin", "all-verifier" ]
```
This was mis-picked from Choi other PR when build on s390x machine, I'll leave it to Choi and using original one.

mkulke reviewed (approved):
lgtm.

I'll suggest to wait for the attester module to be merged first and then bump the kbs/tools/client guest-component dependency revision for e2e tests

Xynnn007 reviewed (approved):
Thanks. lgtm. Let's get this merged after https://github.com/confidential-containers/guest-components/pull/492

huoqifeng commented:
@mkulke @Xynnn007 thank you very much for reviewing the PR and gave so many useful suggestions! thanks for your approval.

mythi commented on kbs/src/api/src/http/attest.rs:
```diff
@@ -17,11 +18,19 @@ pub(crate) async fn auth(
 
-    let session = SessionStatus::auth(request.0, **timeout)
+    let extra_params = attestation_service
+        .generate_supplemental_challenge(request.tee, request.extra_params.clone())
```
we have #242 RFC to suggest how `extra_params` are used. perhaps we can review the new usages in it first and have them documented before any code gets added.

Xynnn007 replied:
Sounds good. The only flaws not covered well in this PR is how KBS interacts with AS. In the beginning we have refactored AS & KBS to [avoid coupling](https://github.com/confidential-containers/trustee/pull/240). So the core idea is that we should not use a structure of KBS as AS's API.

An alternative way is to put the things into `nonce` rather than `extra_params` for IBM SE here. `extra_params` could be decided & appended in the kbs side by the AS plugin type.


btw, I still prefer to integrate IBM SE inside only KBS part rather than also covering AS part. Let's see in future.

Xynnn007 replied:
In this way, the [nonce()](https://github.com/confidential-containers/trustee/pull/163/commits/1e3b172a5c16d357f68799e0b2ed4c0075fdc362#diff-b83a2a4e4135ca8f4193335a1a8d660b5f1dadf86f8cad301b44c4f2a6867a2eR37) API means the same as `generate_supplemental_challenge()` here.

huoqifeng replied:
> `generate_supplemental_challenge`

It's similar for `nonce()` and `generate_supplemental_challenge()`, the difference is: for IBM SE which is generated by verifier as part of the challenge. Do you think reasonable to delegate nonce() generation from KBS to verifier?

huoqifeng replied:
The key point is should AS or KBS generate the attestation challenge. If challenge is specific to verifier or holds verifier specific info, it should be generated by AS/Verifier (IBM SE is such as case).

Xynnn007 replied:
I know that. What I mean are
1. The challenges generated by IBM SE might better to be put to `nonce` rather than `extra_params`
2. `nonce` in @mythi 's PR also aims to be from AS (although the default impl is from KBS), thus the same as `generate_supplemental_challenge`
3. `extra_params` would have some other semantice as #242 proposes. All the contents now is determined by the AS type, but no need to interact with AS to get.
4. Also, as 1stR-A in R-C-A-R are both handled by backend AS, we could rename `generate_supplemental_challenge` in KBS side to `generate_challenge` and directly returns a `Challenge` structure.

Let me write a commit for this.

huoqifeng replied:
The original names is `generate_challenge` :-)

Xynnn007 replied:
See this commit https://github.com/Xynnn007/kbs/commit/1655c266f67cbdb421659366af655e50662cb68f

huoqifeng replied:
> See this commit [Xynnn007@1655c26](https://github.com/Xynnn007/kbs/commit/1655c266f67cbdb421659366af655e50662cb68f)

I agree with this approach, this is actually my original thoughts :-), @Xynnn007 do you want me picking it and revise or do you want get the commit merged firstly?

Xynnn007 replied:
If possible, pick it together with this pr

mythi replied:
LGTM too, seems it also allows to set  per-AS `"supported-hash-algorithms": ["sha256", "sha384", "sha512"]` as suggested by #242.

But perhaps squash the commits a  bit?

huoqifeng replied:
@Xynnn007 I changed the guest-components accordingly also here https://github.com/confidential-containers/guest-components/pull/492/commits/cdeabd46330defd3e401175a1d7b3c4bf7e8e8e3

huoqifeng commented:
Squashed commits.

Xynnn007 reviewed (approved):
lgtm

Xynnn007 commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,311 @@
+#[serde_as]
+#[derive(Debug, Serialize, Deserialize)]
+pub struct SeAttestationClaims {
+    #[serde_as(as = "Base64")]
```
I suggest that we use hex/base16 to encode the parsed claims. This would help users have a common way to set policy. see https://github.com/confidential-containers/trustee/issues/372

huoqifeng replied:
Yeah, I'm happy to keep aligned with other drivers.

steffen-eiden commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,311 @@
+#[derive(Debug)]
+pub struct RealSeVerifier {
+    rsa_private_key: PKey<Private>,
+    rsa_public_key: PKey<Public>,
```
I'd prefer EC keys, but that's up to the projects maintainers.

SE attestation uses EC-keys with spec521r1 curve

huoqifeng replied:
I'm happy to use ECIES approach, after some searching I don't find a handy openssl example, I think pure rust approach is not good for us. Let me add it in my todo. @steffen-eiden WDYT?

steffen-eiden replied:
Instead of reading in a RSA key as pem-file just read in an ec key as pem-file (and rename the variable)
->
`SE_MEASUREMENT_ENCR_KEY_PRIVATE` should contain an EC key in PEm format and you are using EC keys instead of RSA keys.

the PKey abstracts over the exact key type (RSA, EC, HMAC, ....) see [OpenSSL documenatation (C)](https://www.openssl.org/docs/man3.0/man3/EVP_PKEY_new.html)

huoqifeng replied:
oh, you don't mean ECIES but just replace the RSA key to EC, OK.

fitzthum commented on kbs/src/api/src/attestation/mod.rs:
```diff
@@ -32,6 +34,21 @@ pub trait Attest: Send + Sync {
+        Ok(Challenge {
+            nonce,
+            extra_params: String::new(),
+        })
```
Do we really need to implement this method here? It's just a fallback, right?

huoqifeng replied:
Yes, this is a fallback.

fitzthum commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,311 @@
+
+    for entry in fs::read_dir(dir)? {
+        let entry = entry?;
+        let path = entry.path();
```
I think you can probably collapse these two lines into the next one.

huoqifeng replied:
Good point! thank you, fixed.

fitzthum reviewed (commented):
Looks pretty good. A few comments.

huoqifeng commented on kbs/src/api/src/attestation/coco/grpc.rs:
```diff
@@ -118,6 +123,46 @@ impl Attest for GrpcClientPool {
+                    .context("CoCo AS client: serialize tee type failed.")?
+                    .trim_end_matches('"')
+                    .trim_start_matches('"')
+                    .to_string();
```
Good point, thank you Tobin, fixed.

huoqifeng commented on attestation-service/verifier/src/se/mod.rs:
```diff
@@ -0,0 +1,45 @@
+        let se_verifier = ONCE
+            .get_or_try_init(|| async { RealSeVerifier::new() })
+            .await?;
+        warn!("IBM SE does not support initdata.");
```
Yeah, I'll remove it and add a comment here.

fitzthum replied:
See how the AZ SNP verifier [handles it](https://github.com/confidential-containers/trustee/blob/main/attestation-service/verifier/src/az_snp_vtpm/mod.rs#L103-L105)

huoqifeng replied:
Got it, fixed, thank you @fitzthum

huoqifeng commented on attestation-service/docs/parsed_claims.md:
```diff
@@ -93,6 +93,14 @@ The claim inherit the fields from the SEV-SNP claim with and additional `tpm` hi
+## IBM Secure Execution (SE)
+- `se.version`: The version this quote structure.
+- `se.cuid`: The config uid.
+- `se.hdr.tag`: SE header tag (seht)
```
Good point, thank you Steffen!

huoqifeng commented on attestation-service/docs/parsed_claims.md:
```diff
@@ -93,6 +93,14 @@ The claim inherit the fields from the SEV-SNP claim with and additional `tpm` hi
+- `se.hdr.tag`: SE header tag (seht)
+- `se.image.phkh`: SE image public host key hash
+- `se.attestation.phkh`: SE attestation public host key hash
+- `se.user_data`: Custom attestation key owner data.
```
Good point, thank you Steffen!

huoqifeng commented on attestation-service/docs/parsed_claims.md:
```diff
@@ -93,6 +93,14 @@ The claim inherit the fields from the SEV-SNP claim with and additional `tpm` hi
 
+## IBM Secure Execution (SE)
+- `se.version`: The version this quote structure.
+- `se.cuid`: The config uid.
```
Good point, thank you Steffen.

huoqifeng commented on attestation-service/verifier/src/lib.rs:
```diff
@@ -152,6 +165,17 @@ pub trait Verifier {
+    ///
+    /// A optional `tee_parameters` comes from the attester side as the input.
+    async fn generate_supplemental_challenge(&self, _tee_parameters: String) -> Result<String> {
+        Ok(String::new())
```
It's the default implementation as a fallback.

huoqifeng commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,311 @@
+        if !measurement.eq_secure(&se_response.measurement) {
+            debug!("Recieved: {:?}", se_response.measurement);
+            debug!("Calculated: {:?}", measurement.as_ref());
+            warn!("Attestation measurement verification failed. Calculated and received attestation measurement are not equal.");
```
Yes, agreed, fixed.

huoqifeng commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,311 @@
+            bail!("Failed to verify the measurement!");
+        }
+
+        // TODO check self.user_data.image_btph with previous saved value
```
Removed.

huoqifeng commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,311 @@
+            .attestation_public_host_key_hash()
+            .ok_or(anyhow!("Failed to get attestation_public_host_key_hash."))?;
+
+        // TODO image_phkh and attestation_phkh with previous saved value
```
Removed

huoqifeng commented on attestation-service/verifier/src/se/mod.rs:
```diff
@@ -0,0 +1,45 @@
+
+pub mod ibmse;
+
+static ONCE: OnceCell<RealSeVerifier> = OnceCell::const_new();
```
Good point, fixed.

fitzthum replied:
I mean `ONCE` -> `VERIFIER`.

huoqifeng replied:
oh, got it, fixed, thank you @fitzthum

huoqifeng commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,311 @@
+            bail!("The nonce vector must have exactly 16 elements.");
+        }
+
+        let nonce_array: [u8; 16] = nonce
```
Merged these 2.

Xynnn007 commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,311 @@
+
+#[serde_as]
+#[derive(Debug, Serialize, Deserialize)]
+pub struct SeAttestationResponse {
```
For other archs, there should be a 3rd crate define this. For SE, `pv_s390x` might be a good choice.

huoqifeng replied:
Yeah, pv_s390x is an option, another option is to create a repo under IBM org. Can we do this refine later? Which should be se only change. While in this PR, there is some AS/KBS API change, I'd like to separate. @steffen-eiden @Xynnn007 WDYT?

steffen-eiden replied:
you can always refactor later IMO - if you want move fast with this PR i don't see an issue here.

if you meant with `pv_s390x` the [pv crate](https://docs.rs/s390_pv/0.10.0/s390_pv/) I have to say no.
This crate should not contain any implementation details about the attestation.

huoqifeng commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,311 @@
+
+        // TODO check self.user_data.image_btph with previous saved value
+
+        let mut att_flags = AttestationFlags::default();
```
@steffen-eiden the flags bind to the claims, we'll have to change claims format if flags change. Most likely we'll use fixed flags even more options provided. We need change flags and claims at same time at least. Can we leave this change to later?

steffen-eiden replied:
Yes, IMO we can to that later if more flags and algorithms are supported by the s390 attestation flow.
IDK anything about the compatibility claims of this project so just make sure you do not block further extensions to the claims, request, and response structs.

-- mark as resolved --

huoqifeng commented on attestation-service/verifier/src/se/ibmse.rs:
```diff
@@ -0,0 +1,311 @@
+        );
+
+        let measurement =
+            AttestationMeasurement::calculate(items, AttestationMeasAlg::HmacSha512, &meas_key)?;
```
@steffen-eiden same as flags, I'd like to leave this as later change, wdyt?

steffen-eiden replied:
same as above

-- mark as resolved --

fitzthum reviewed (approved):
LGTM

I do want to emphasize that since this verifier does not check the report data, it is not secure. Hopefully that will be addressed in a follow-up. I think this PR is fine for getting all the support code in.

huoqifeng commented:
> LGTM
> 
> I do want to emphasize that since this verifier does not check the report data, it is not secure. Hopefully that will be addressed in a follow-up. I think this PR is fine for getting all the support code in.

Thank you! @fitzthum , yes, the checking algorithm checks the evidence from TEE guest and re-calculated value from verifier, which also based on same SE OS Image. We'll add additional check in a follow up PR.

Xynnn007 commented:
https://github.com/confidential-containers/guest-components/pull/492 was merged. we should update the rev in https://github.com/confidential-containers/trustee/blob/main/kbs/tools/client/Cargo.toml#L21
